<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pbcore.io.align package &mdash; pbcore 1.2.1 documentation</title>
    
    <link rel="stylesheet" href="_static/pacbio.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.2.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="pbcore 1.2.1 documentation" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pbcore 1.2.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="pbcore-io-align-package">
<h1>pbcore.io.align package<a class="headerlink" href="#pbcore-io-align-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-pbcore.io.align.BamAlignment">
<span id="pbcore-io-align-bamalignment-module"></span><h2>pbcore.io.align.BamAlignment module<a class="headerlink" href="#module-pbcore.io.align.BamAlignment" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pbcore.io.align.BamAlignment.BamAlignment">
<em class="property">class </em><code class="descclassname">pbcore.io.align.BamAlignment.</code><code class="descname">BamAlignment</code><span class="sig-paren">(</span><em>bamReader</em>, <em>pysamAlignedRead</em>, <em>rowNumber=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.align._AlignmentMixin.AlignmentRecordMixin</span></code></p>
<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.DeletionQV">
<code class="descname">DeletionQV</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.DeletionQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.DeletionTag">
<code class="descname">DeletionTag</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.DeletionTag" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.HoleNumber">
<code class="descname">HoleNumber</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.HoleNumber" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.IPD">
<code class="descname">IPD</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.IPD" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.InsertionQV">
<code class="descname">InsertionQV</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.InsertionQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.MapQV">
<code class="descname">MapQV</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.MapQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.MergeQV">
<code class="descname">MergeQV</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.MergeQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.PulseWidth">
<code class="descname">PulseWidth</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.PulseWidth" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.SubstitutionQV">
<code class="descname">SubstitutionQV</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.SubstitutionQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.barcode">
<code class="descname">barcode</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.barcode" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.barcodeName">
<code class="descname">barcodeName</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.barcodeName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.clippedTo">
<code class="descname">clippedTo</code><span class="sig-paren">(</span><em>bamAln</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.clippedTo" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new <cite>BamAlignment</cite> that refers to a subalignment of
this alignment, as induced by clipping to reference
coordinates <cite>refStart</cite> to <cite>refEnd</cite>.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This function takes time linear in the length of the alignment.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.identity">
<code class="descname">identity</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.identity" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.isCCS">
<code class="descname">isCCS</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.isCCS" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.isForwardStrand">
<code class="descname">isForwardStrand</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.isForwardStrand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.isMapped">
<code class="descname">isMapped</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.isMapped" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.isReverseStrand">
<code class="descname">isReverseStrand</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.isReverseStrand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.isUnmapped">
<code class="descname">isUnmapped</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.isUnmapped" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.mapQV">
<code class="descname">mapQV</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.mapQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.movieName">
<code class="descname">movieName</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.movieName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.numPasses">
<code class="descname">numPasses</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.numPasses" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.pulseFeature">
<code class="descname">pulseFeature</code><span class="sig-paren">(</span><em>featureName</em>, <em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.pulseFeature" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve the pulse feature as indicated.
- <cite>aligned</cite>    : whether gaps should be inserted to reflect the alignment
- <cite>orientation</cite>: &#8220;native&#8221; or &#8220;genomic&#8221;</p>
<p>Note that this function assumes the the feature is stored in
native orientation in the file, so it is not appropriate to
use this method to fetch the read or the qual, which are
oriented genomically in the file.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.qEnd">
<code class="descname">qEnd</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.qEnd" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.qId">
<code class="descname">qId</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.qId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.qLen">
<code class="descname">qLen</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.qLen" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.qName">
<code class="descname">qName</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.qName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.qStart">
<code class="descname">qStart</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.qStart" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.queryEnd">
<code class="descname">queryEnd</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.queryEnd" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.queryName">
<code class="descname">queryName</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.queryName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.queryStart">
<code class="descname">queryStart</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.queryStart" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.read">
<code class="descname">read</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.read" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.readGroupInfo">
<code class="descname">readGroupInfo</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.readGroupInfo" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.readPositions">
<code class="descname">readPositions</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.readPositions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array of read positions.</p>
<p>If aligned is True, the array has the same length as the
alignment and readPositions[i] = read position of the i&#8217;th
column in the oriented alignment.</p>
<p>If aligned is False, the array has the same length as the
mapped reference segment and readPositions[i] = read position
of the i&#8217;th base in the oriented reference segment.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.readType">
<code class="descname">readType</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.readType" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.reader">
<code class="descname">reader</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.reader" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.reference">
<code class="descname">reference</code><span class="sig-paren">(</span><em>bamAln</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.reference" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.referenceId">
<code class="descname">referenceId</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.referenceId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.referenceInfo">
<code class="descname">referenceInfo</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.referenceInfo" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.referenceName">
<code class="descname">referenceName</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.referenceName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.referencePositions">
<code class="descname">referencePositions</code><span class="sig-paren">(</span><em>bamAln</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.referencePositions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array of reference positions.</p>
<p>If aligned is True, the array has the same length as the
alignment and referencePositions[i] = reference position of
the i&#8217;th column in the oriented alignment.</p>
<p>If aligned is False, the array has the same length as the read
and referencePositions[i] = reference position of the i&#8217;th
base in the oriented read.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.sequencingChemistry">
<code class="descname">sequencingChemistry</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.sequencingChemistry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.tId">
<code class="descname">tId</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.tId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.transcript">
<code class="descname">transcript</code><span class="sig-paren">(</span><em>orientation='native'</em>, <em>style='gusfield'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>A text representation of the alignment moves (see Gusfield).
This can be useful in pretty-printing an alignment.</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.unrolledCigar">
<code class="descname">unrolledCigar</code><span class="sig-paren">(</span><em>bamAln</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.unrolledCigar" title="Permalink to this definition">¶</a></dt>
<dd><p>Run-length decode the CIGAR encoding, and orient.  Clipping ops are removed.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamAlignment.BamAlignment.zScore">
<code class="descname">zScore</code><a class="headerlink" href="#pbcore.io.align.BamAlignment.BamAlignment.zScore" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pbcore.io.align.BamIO">
<span id="pbcore-io-align-bamio-module"></span><h2>pbcore.io.align.BamIO module<a class="headerlink" href="#module-pbcore.io.align.BamIO" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pbcore.io.align.BamIO.BamReader">
<em class="property">class </em><code class="descclassname">pbcore.io.align.BamIO.</code><code class="descname">BamReader</code><span class="sig-paren">(</span><em>fname</em>, <em>referenceFastaFname=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamIO.BamReader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.align.BamIO._BamReaderBase</span></code>, <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.align._AlignmentMixin.AlignmentReaderMixin</span></code></p>
<p>Reader for a BAM with a bam.bai (SAMtools) index, but not a
bam.pbi (PacBio) index.  Supports basic BAM operations.</p>
<dl class="attribute">
<dt id="pbcore.io.align.BamIO.BamReader.index">
<code class="descname">index</code><a class="headerlink" href="#pbcore.io.align.BamIO.BamReader.index" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamIO.BamReader.readsInRange">
<code class="descname">readsInRange</code><span class="sig-paren">(</span><em>winId</em>, <em>winStart</em>, <em>winEnd</em>, <em>justIndices=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamIO.BamReader.readsInRange" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.align.BamIO.IndexedBamReader">
<em class="property">class </em><code class="descclassname">pbcore.io.align.BamIO.</code><code class="descname">IndexedBamReader</code><span class="sig-paren">(</span><em>fname</em>, <em>referenceFastaFname=None</em>, <em>sharedIndex=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamIO.IndexedBamReader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.align.BamIO._BamReaderBase</span></code>, <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.align._AlignmentMixin.IndexedAlignmentReaderMixin</span></code></p>
<p>A <cite>IndexedBamReader</cite> is a BAM reader class that uses the
<code class="docutils literal"><span class="pre">bam.pbi</span></code> (PacBio BAM index) file to enable random access by
&#8220;row number&#8221; and to provide access to precomputed semantic
information about the BAM records</p>
<dl class="method">
<dt id="pbcore.io.align.BamIO.IndexedBamReader.atRowNumber">
<code class="descname">atRowNumber</code><span class="sig-paren">(</span><em>rn</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamIO.IndexedBamReader.atRowNumber" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.BamIO.IndexedBamReader.index">
<code class="descname">index</code><a class="headerlink" href="#pbcore.io.align.BamIO.IndexedBamReader.index" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.BamIO.IndexedBamReader.readsInRange">
<code class="descname">readsInRange</code><span class="sig-paren">(</span><em>winId</em>, <em>winStart</em>, <em>winEnd</em>, <em>justIndices=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BamIO.IndexedBamReader.readsInRange" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pbcore.io.align.BlasrIO">
<span id="pbcore-io-align-blasrio-module"></span><h2>pbcore.io.align.BlasrIO module<a class="headerlink" href="#module-pbcore.io.align.BlasrIO" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pbcore.io.align.BlasrIO.M4Record">
<em class="property">class </em><code class="descclassname">pbcore.io.align.BlasrIO.</code><code class="descname">M4Record</code><a class="headerlink" href="#pbcore.io.align.BlasrIO.M4Record" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>Record for alignment summary record output from BLASR -m 4 option</p>
<dl class="classmethod">
<dt id="pbcore.io.align.BlasrIO.M4Record.fromString">
<em class="property">classmethod </em><code class="descname">fromString</code><span class="sig-paren">(</span><em>s</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BlasrIO.M4Record.fromString" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.align.BlasrIO.M4Reader">
<em class="property">class </em><code class="descclassname">pbcore.io.align.BlasrIO.</code><code class="descname">M4Reader</code><span class="sig-paren">(</span><em>f</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BlasrIO.M4Reader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.base.ReaderBase</span></code></p>
<p>Reader for -m 4 formatted alignment summary information from BLASR</p>
</dd></dl>

<dl class="class">
<dt id="pbcore.io.align.BlasrIO.M5Record">
<em class="property">class </em><code class="descclassname">pbcore.io.align.BlasrIO.</code><code class="descname">M5Record</code><a class="headerlink" href="#pbcore.io.align.BlasrIO.M5Record" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>Record for alignment summary record output from BLASR -m 5 option</p>
<dl class="classmethod">
<dt id="pbcore.io.align.BlasrIO.M5Record.fromString">
<em class="property">classmethod </em><code class="descname">fromString</code><span class="sig-paren">(</span><em>s</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BlasrIO.M5Record.fromString" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.align.BlasrIO.M5Reader">
<em class="property">class </em><code class="descclassname">pbcore.io.align.BlasrIO.</code><code class="descname">M5Reader</code><span class="sig-paren">(</span><em>f</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.BlasrIO.M5Reader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.base.ReaderBase</span></code></p>
<p>Reader for -m 5 formatted alignment summary information from BLASR</p>
</dd></dl>

</div>
<div class="section" id="module-pbcore.io.align.CmpH5IO">
<span id="pbcore-io-align-cmph5io-module"></span><h2>pbcore.io.align.CmpH5IO module<a class="headerlink" href="#module-pbcore.io.align.CmpH5IO" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader">
<em class="property">class </em><code class="descclassname">pbcore.io.align.CmpH5IO.</code><code class="descname">CmpH5Reader</code><span class="sig-paren">(</span><em>filenameOrH5File</em>, <em>sharedIndex=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.base.ReaderBase</span></code>, <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.align._AlignmentMixin.IndexedAlignmentReaderMixin</span></code></p>
<p>The <cite>CmpH5Reader</cite> class is a lightweight and efficient API for
accessing PacBio <code class="docutils literal"><span class="pre">cmp.h5</span></code> alignment files.  Alignment records
can be obtained via random access (via Python indexing/slicing),
iteration, or range queries (via readsInRange).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pbcore.data</span>                <span class="c"># For an example data file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pbcore.io</span> <span class="kn">import</span> <span class="n">CmpH5Reader</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filename</span> <span class="o">=</span> <span class="n">pbcore</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">getCmpH5</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">CmpH5Reader</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">CmpH5 alignment: -    1          0        290</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>  
<span class="go">[CmpH5 alignment: -    1          0        290,</span>
<span class="go"> CmpH5 alignment: +    1          0        365]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sum</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">readLength</span> <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="n">c</span><span class="p">)</span>
<span class="go">26103</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.ReadGroupID">
<code class="descname">ReadGroupID</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.ReadGroupID" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.alignmentIndex">
<code class="descname">alignmentIndex</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.alignmentIndex" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the alignment index data structure, which is the
central data structure in the cmp.h5 file, as a numpy
recarray.</p>
<p>The <cite>dtype</cite> of the recarray is:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">dtype</span><span class="p">([(</span><span class="s">&#39;AlnID&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;AlnGroupID&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;MovieID&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;RefGroupID&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;tStart&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;tEnd&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;RCRefStrand&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;HoleNumber&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;SetNumber&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;StrobeNumber&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;MoleculeID&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;rStart&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;rEnd&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;MapQV&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;nM&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;nMM&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;nIns&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;nDel&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;Offset_begin&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;Offset_end&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;nBackRead&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;nReadOverlap&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)])</span>
</pre></div>
</div>
<p>Access to the alignment index is provided to allow users to
perform vectorized computations over all alignments in the file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">alignmentIndex</span><span class="o">.</span><span class="n">MapQV</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>
<span class="go">array([254, 254,   0, 254, 254, 254, 254, 254, 254, 254], dtype=uint32)</span>
</pre></div>
</div>
<p>Alignment index fields are also exposed as fields of the
<cite>CmpH5Reader</cite> object, allowing a convenient shorthand.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">MapQV</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>
<span class="go">array([254, 254,   0, 254, 254, 254, 254, 254, 254, 254], dtype=uint32)</span>
</pre></div>
</div>
<p>The alignment index row for a given alignment can also be
accessed directly as a field of a <cite>CmpH5Alignment</cite> object</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="p">[</span><span class="mi">26</span><span class="p">]</span><span class="o">.</span><span class="n">MapQV</span>
<span class="go">254</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.barcode">
<code class="descname">barcode</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.barcode" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dict mapping of barcode name to integer barcode.
Behavior undefined if file is not barcoded.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.barcodeName">
<code class="descname">barcodeName</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.barcodeName" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dict mapping of barcode integer id to name.
Behavior undefined if file is not barcoded.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.barcodes">
<code class="descname">barcodes</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.barcodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array of barcode integer ids, of the same length as the
alignment array.</p>
<p>Behavior undefined if file is not barcoded.</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.close">
<code class="descname">close</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.close" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.hasPulseFeature">
<code class="descname">hasPulseFeature</code><span class="sig-paren">(</span><em>featureName</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.hasPulseFeature" title="Permalink to this definition">¶</a></dt>
<dd><p>Are the datasets for pulse feature <cite>featureName</cite> loaded in
this file?  Specifically, is it loaded for all movies within
this cmp.h5?</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">hasPulseFeature</span><span class="p">(</span><span class="s">&quot;InsertionQV&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">hasPulseFeature</span><span class="p">(</span><span class="s">&quot;MergeQV&quot;</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.holeNumber">
<code class="descname">holeNumber</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.holeNumber" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.index">
<code class="descname">index</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.index" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.isBarcoded">
<code class="descname">isBarcoded</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.isBarcoded" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.isEmpty">
<code class="descname">isEmpty</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.isEmpty" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.isSorted">
<code class="descname">isSorted</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.isSorted" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.mapQV">
<code class="descname">mapQV</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.mapQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.movieInfo">
<code class="descname">movieInfo</code><span class="sig-paren">(</span><em>movieId</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.movieInfo" title="Permalink to this definition">¶</a></dt>
<dd><div class="deprecated">
<p><span class="versionmodified">Deprecated since version 0.9.2: </span>Use <a class="reference internal" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupInfo" title="pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupInfo"><code class="xref py py-attr docutils literal"><span class="pre">readGroupInfo</span></code></a>, which is compatible with BAM usage</p>
</div>
<p>Access information about a movie whose reads are represented
in the file.</p>
<p>The returned value is a record from the <a class="reference internal" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.movieInfoTable" title="pbcore.io.align.CmpH5IO.CmpH5Reader.movieInfoTable"><code class="xref py py-attr docutils literal"><span class="pre">movieInfoTable</span></code></a></p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.movieInfoTable">
<code class="descname">movieInfoTable</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.movieInfoTable" title="Permalink to this definition">¶</a></dt>
<dd><div class="deprecated">
<p><span class="versionmodified">Deprecated since version 0.9.2: </span>Use <a class="reference internal" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupTable" title="pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupTable"><code class="xref py py-attr docutils literal"><span class="pre">readGroupTable</span></code></a>, which is compatible with BAM usage</p>
</div>
<p>Return a numpy recarray summarizing source movies for the
reads in this file.</p>
<p>The <cite>dtype</cite> of this recarray is:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">dtype</span><span class="p">([(</span><span class="s">&#39;ID&#39;</span><span class="p">,</span> <span class="s">&#39;int&#39;</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;Name&#39;</span><span class="p">,</span> <span class="s">&#39;string&#39;</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;FrameRate&#39;</span><span class="p">,</span> <span class="s">&#39;float&#39;</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;TimeScale&#39;</span><span class="p">,</span> <span class="s">&#39;float&#39;</span><span class="p">)])</span>
</pre></div>
</div>
<p><cite>TimeScale</cite> is the factor to multiply time values (IPD,
PulseWidth) by in order to get times in seconds.  The
<cite>FrameRate</cite> field should <em>not</em> be used directly as it will be
NaN for pre-1.3 cmp.h5 files.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.movieNames">
<code class="descname">movieNames</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.movieNames" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.pulseFeaturesAvailable">
<code class="descname">pulseFeaturesAvailable</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.pulseFeaturesAvailable" title="Permalink to this definition">¶</a></dt>
<dd><p>What pulse features are available in this cmp.h5 file?</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">pulseFeaturesAvailable</span><span class="p">()</span>
<span class="go">[u&#39;QualityValue&#39;, u&#39;IPD&#39;, u&#39;PulseWidth&#39;, u&#39;InsertionQV&#39;, u&#39;DeletionQV&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.qId">
<code class="descname">qId</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.qId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupInfo">
<code class="descname">readGroupInfo</code><span class="sig-paren">(</span><em>rgId</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Access information about a movie whose reads are represented
in the file.</p>
<p>The returned value is a record from the <a class="reference internal" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupTable" title="pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupTable"><code class="xref py py-attr docutils literal"><span class="pre">readGroupTable</span></code></a></p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupTable">
<code class="descname">readGroupTable</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.readGroupTable" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.readType">
<code class="descname">readType</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.readType" title="Permalink to this definition">¶</a></dt>
<dd><p>Either &#8220;standard&#8221; or &#8220;CCS&#8221;, indicating the type of reads that
were aligned to the reference.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">readType</span>
<span class="go">&#39;standard&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.readsInRange">
<code class="descname">readsInRange</code><span class="sig-paren">(</span><em>refKey</em>, <em>refStart</em>, <em>refEnd</em>, <em>justIndices=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.readsInRange" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of reads overlapping (i.e., intersecting&#8212;not
necessarily spanning) a given reference window.</p>
<p>If <cite>justIndices</cite> is <code class="docutils literal"><span class="pre">False</span></code>, the list returned will contain
<cite>CmpH5Alignment</cite> objects.</p>
<p>If <cite>justIndices</cite> is <code class="docutils literal"><span class="pre">True</span></code>, the list returned will contain
row numbers in the alignment index table.  Slicing the
<cite>CmpH5Reader</cite> object with these row numbers can be used to get
the corresponding <cite>CmpH5Alignment</cite> objects.</p>
<p>The contig key can be either the <code class="docutils literal"><span class="pre">RefID</span></code>, or the short name
(FASTA header up to first space).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">readsInRange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span> 
<span class="go">[CmpH5 alignment: -    1          0        290,</span>
<span class="go"> CmpH5 alignment: +    1          0        365]</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">rowNumbers</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">readsInRange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">justIndices</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowNumbers</span>
<span class="go">array([0, 1], dtype=uint32)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.referenceInfo">
<code class="descname">referenceInfo</code><span class="sig-paren">(</span><em>key</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.referenceInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Access information about a reference that was aligned against.
Key can be reference ID (integer), name (&#8220;ref000001&#8221;), full
name (e.g. &#8220;lambda_NEB3011&#8221;), truncated full name (full name
up to the first whitespace, following the samtools convention)
or MD5 sum hex string (e.g. &#8220;a1319ff90e994c8190a4fe6569d0822a&#8221;).</p>
<p>The returned value is a record from the :ref:referenceInfoTable .</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ri</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">referenceInfo</span><span class="p">(</span><span class="s">&quot;ref000001&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ri</span><span class="o">.</span><span class="n">FullName</span>
<span class="go">&#39;lambda_NEB3011&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ri</span><span class="o">.</span><span class="n">MD5</span>
<span class="go">&#39;a1319ff90e994c8190a4fe6569d0822a&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.referenceInfoTable">
<code class="descname">referenceInfoTable</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.referenceInfoTable" title="Permalink to this definition">¶</a></dt>
<dd><p id="referenceinfotable">Return a numpy recarray summarizing the references that were
aligned against.</p>
<p>The <cite>dtype</cite> of this recarray is:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">dtype</span><span class="p">([(</span><span class="s">&#39;RefInfoID&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;ID&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;Name&#39;</span><span class="p">,</span> <span class="n">string</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;FullName&#39;</span><span class="p">,</span> <span class="n">string</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;Length&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;MD5&#39;</span><span class="p">,</span> <span class="n">string</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;StartRow&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span>
       <span class="p">(</span><span class="s">&#39;EndRow&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="p">])</span>
</pre></div>
</div>
<p>(the last two columns are omitted for unsorted <cite>cmp.h5</cite> files).</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.sequencingChemistry">
<code class="descname">sequencingChemistry</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.sequencingChemistry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.softwareVersion">
<code class="descname">softwareVersion</code><span class="sig-paren">(</span><em>programName</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.softwareVersion" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the version of program <cite>programName</cite> that processed
this file.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.version">
<code class="descname">version</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.version" title="Permalink to this definition">¶</a></dt>
<dd><p>The CmpH5 format version string.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">version</span>
<span class="go">&#39;1.2.0.SF&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Reader.versionAtLeast">
<code class="descname">versionAtLeast</code><span class="sig-paren">(</span><em>minimalVersion</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Reader.versionAtLeast" title="Permalink to this definition">¶</a></dt>
<dd><p>Compare the file version to <cite>minimalVersion</cite>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">versionAtLeast</span><span class="p">(</span><span class="s">&quot;1.3.0&quot;</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment">
<em class="property">class </em><code class="descclassname">pbcore.io.align.CmpH5IO.</code><code class="descname">CmpH5Alignment</code><span class="sig-paren">(</span><em>cmph5</em>, <em>rowNumber</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">pbcore.io.align._AlignmentMixin.AlignmentRecordMixin</span></code></p>
<p>A lightweight class representing a single alignment record in a
CmpH5 file, providing access to all columns of a single row of the
alignment index, and on-demand access to the corresponding
sequence and pulse features.</p>
<p><cite>CmpH5Alignment</cite> objects are obtained by slicing a
<cite>CmpH5Reader</cite> object:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="p">[</span><span class="mi">26</span><span class="p">]</span>
<span class="go">CmpH5 alignment: +    1       7441       7699</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">c</span><span class="p">[</span><span class="mi">26</span><span class="p">]</span>
<span class="go">CmpH5 alignment: +    1       7441       7699</span>

<span class="go">Read:        m110818_075520_42141_c100129202555500000315043109121112_s2_p0/1009/44_322</span>
<span class="go">Reference:   lambda_NEB3011</span>

<span class="go">Read length: 278</span>
<span class="go">Concordance: 0.871</span>

<span class="go">  12345678901234567890123456789001223456789012345678890112345678990112344567890011</span>
<span class="go">  AACTGGTCACGGTCGTGGCACTGGTGAAG-CT-GCATACTGATGCACTT-CAC-GCCACGCG-GG-ATG-AACCTG-T-G</span>
<span class="go">  |||||||  ||||  ||||||||| |||| || ||||||||| |||||  ||| |||||||| || ||| |||||| | |</span>
<span class="go">  AACTGGT--CGGT--TGGCACTGG-GAAGCCTTGCATACTGA-GCACT-CCACGGCCACGCGGGGAATGAAACCTGGTGG</span>


<span class="go">  23456789012345678900123456678901234456789012345678901234566789011234556789012345</span>
<span class="go">  GCATTTGTGCTGCCGGGA-ACGGCG-TTTCGTGT-CTCTGCCGGTGTGGCAGCCGAA-ATGAC-AGAG-CGCGGCCTGGC</span>
<span class="go">  |||||||||||||||||| |||||| |||||| | |||||||||||||||||||||| ||||| |||| |||||||||||</span>
<span class="go">  GCATTTGTGCTGCCGGGAAACGGCGTTTTCGT-TCCTCTGCCGGTGTGGCAGCCGAAAATGACAAGAGCCGCGGCCTGGC</span>


<span class="go">  67899012345677890123456789901123456789901233456789012345678901234556778901223455</span>
<span class="go">  CAG-AATGCAAT-AACGGGAGGCGC-TG-TGGCTGAT-TTCG-ATAACCTGTTCGATGCTGCCAT-TG-CCCGC-GCC-G</span>
<span class="go">  ||| |||||||| |||||||||||| || |||||||| |||| |||||||||||||||||||||| || ||||| ||| |</span>
<span class="go">  CAGAAATGCAATAAACGGGAGGCGCTTGCTGGCTGATTTTCGAATAACCTGTTCGATGCTGCCATTTGACCCGCGGCCGG</span>


<span class="go">  6678901234567889012345667890123456789012345678</span>
<span class="go">  -ATGAAACGATAC-GCGGGTAC-ATGGGAACGTCAGCCACCATTAC</span>
<span class="go">   |||||||||||| |||||||| |||||||||||||||||||||||</span>
<span class="go">  AATGAAACGATACGGCGGGTACAATGGGAACGTCAGCCACCATTAC</span>
</pre></div>
</div>
<p>The <cite>orientation</cite> argument to some data access methods determines
how reverse-strand alignments are returned to the user.  <cite>.cmp.h5</cite>
files natively encode reverse strand reads in read-order,
uncomplemented, with the <em>reference</em> reverse-complemented.  Most
analysis applications will want to use the data in this order,
which we term the <em>NATIVE</em> orientation.</p>
<p>Some applications that involve collating or displaying the reads
aligned to the reference genome want the reference to be presented
in its genomic order, and the <em>read</em> to be reverse-complemented.
We term this <em>GENOMIC</em> orientation.</p>
<p>Methods prefixed with <em>aligned</em> return data (bases or features)
that include gaps, which are encoded according to the data type.
Methods not prefixed with <em>aligned</em> excise the gaps.</p>
<p>Sequence and pulse features are not cached.</p>
<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.DeletionQV">
<code class="descname">DeletionQV</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.DeletionQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.DeletionTag">
<code class="descname">DeletionTag</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.DeletionTag" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.IPD">
<code class="descname">IPD</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.IPD" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.InsertionQV">
<code class="descname">InsertionQV</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.InsertionQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.MergeQV">
<code class="descname">MergeQV</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.MergeQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.PulseWidth">
<code class="descname">PulseWidth</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.PulseWidth" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.QualityValue">
<code class="descname">QualityValue</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.QualityValue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.ReadGroupID">
<code class="descname">ReadGroupID</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.ReadGroupID" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.SubstitutionQV">
<code class="descname">SubstitutionQV</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.SubstitutionQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.aEnd">
<code class="descname">aEnd</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.aEnd" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.aStart">
<code class="descname">aStart</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.aStart" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.accuracy">
<code class="descname">accuracy</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.accuracy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the identity of this alignment, calculated as
(#matchs / read length)</p>
<div class="deprecated">
<p><span class="versionmodified">Deprecated since version 0.9.5: </span>Use <a class="reference internal" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.identity" title="pbcore.io.align.CmpH5IO.CmpH5Alignment.identity"><code class="xref py py-attr docutils literal"><span class="pre">identity</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.alignmentArray">
<code class="descname">alignmentArray</code><span class="sig-paren">(</span><em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.alignmentArray" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct access to the raw, encoded aligment array, which is a
packed representation of the aligned read and reference.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.barcode">
<code class="descname">barcode</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.barcode" title="Permalink to this definition">¶</a></dt>
<dd><p>The barcode ID (integer key) for this alignment&#8217;s read
Behavior undefined if file is not barcoded.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.barcodeName">
<code class="descname">barcodeName</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.barcodeName" title="Permalink to this definition">¶</a></dt>
<dd><p>The barcode name (string) for this alignment&#8217;s read
Behavior undefined if file is not barcoded.</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.clippedTo">
<code class="descname">clippedTo</code><span class="sig-paren">(</span><em>refStart</em>, <em>refEnd</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.clippedTo" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new <cite>CmpH5Alignment</cite> that refers to a subalignment of
this alignment, as induced by clipping to reference
coordinates <cite>refStart</cite> to <cite>refEnd</cite>.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This function takes time linear in the length of the alignment.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.cmpH5">
<code class="descname">cmpH5</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.cmpH5" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.holeNumber">
<code class="descname">holeNumber</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.holeNumber" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.identity">
<code class="descname">identity</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.identity" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the identity of this alignment, calculated as
(#matchs / read length)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="p">[</span><span class="mi">26</span><span class="p">]</span><span class="o">.</span><span class="n">identity</span>
<span class="go">0.87050359712230219</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.isForwardStrand">
<code class="descname">isForwardStrand</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.isForwardStrand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.isReverseStrand">
<code class="descname">isReverseStrand</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.isReverseStrand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.mapQV">
<code class="descname">mapQV</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.mapQV" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.movieInfo">
<code class="descname">movieInfo</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.movieInfo" title="Permalink to this definition">¶</a></dt>
<dd><div class="deprecated">
<p><span class="versionmodified">Deprecated since version 0.9.2: </span>Use <a class="reference internal" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.readGroupInfo" title="pbcore.io.align.CmpH5IO.CmpH5Alignment.readGroupInfo"><code class="xref py py-attr docutils literal"><span class="pre">readGroupInfo</span></code></a>, which is compatible with BAM usage</p>
</div>
<p>Returns a record (extracted from the cmph5&#8217;s <cite>movieInfoTable</cite>)
containing information about the movie that the read was
extracted from.  This record should be accessed using
dot-notation, according to the column names documented in
<cite>movieInfoTable</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.movieName">
<code class="descname">movieName</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.movieName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.numPasses">
<code class="descname">numPasses</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.numPasses" title="Permalink to this definition">¶</a></dt>
<dd><p>(CCS only) The number subreads that were used to produce this CCS read.</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.pulseFeature">
<code class="descname">pulseFeature</code><span class="sig-paren">(</span><em>featureName</em>, <em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.pulseFeature" title="Permalink to this definition">¶</a></dt>
<dd><p>Access a pulse feature by name.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.qId">
<code class="descname">qId</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.qId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.read">
<code class="descname">read</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.read" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the read portion of the alignment as a string.</p>
<p>If <cite>aligned</cite> is true, the aligned representation is returned,
including gaps; otherwise the unaligned read basecalls are
returned.</p>
<p>If <cite>orientation</cite> is &#8220;native&#8221;, the returned read bases are
presented in the order they were read by the sequencing
machine.  If <cite>orientation</cite> is &#8220;genomic&#8221;, the returned read
bases are presented in such a way as to collate with the
forward strand of the reference&#8212;which requires reverse
complementation of reverse-strand reads.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.readGroupInfo">
<code class="descname">readGroupInfo</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.readGroupInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the corresponding record from the <cite>readGroupTable</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.readPositions">
<code class="descname">readPositions</code><span class="sig-paren">(</span><em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.readPositions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array of read positions such that
readPositions[i] = read position of the i&#8217;th column
in the alignment.  Insertions are grouped with the following
read base, in the specified orientation.</p>
<p>Length of output array = length of alignment</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.readType">
<code class="descname">readType</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.readType" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.reader">
<code class="descname">reader</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.reader" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.reference">
<code class="descname">reference</code><span class="sig-paren">(</span><em>aligned=True</em>, <em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.reference" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the read portion of the alignment as a string.</p>
<p>If <cite>aligned</cite> is true, the aligned representation of the
reference is returned, including gaps; otherwise the unaligned
reference bases are returned.</p>
<p>If <cite>orientation</cite> is &#8220;native&#8221;, the reference is presented in
the order it is stored in the cmp.h5 file&#8212;for reverse-strand
reads, the reference is reverse-complemented.  If
<cite>orientation</cite> is &#8220;genomic&#8221;, the forward strand reference is returned.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.referenceId">
<code class="descname">referenceId</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.referenceId" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.referenceInfo">
<code class="descname">referenceInfo</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.referenceInfo" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.referenceName">
<code class="descname">referenceName</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.referenceName" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.referencePositions">
<code class="descname">referencePositions</code><span class="sig-paren">(</span><em>orientation='native'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.referencePositions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array of reference positions such that
referencePositions[i] = reference position of the i&#8217;th column
in the alignment.  Insertions are grouped with the following
reference base, in the specified orientation.</p>
<p>Length of output array = length of alignment</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.rowNumber">
<code class="descname">rowNumber</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.rowNumber" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.sequencingChemistry">
<code class="descname">sequencingChemistry</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.sequencingChemistry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.similarity">
<code class="descname">similarity</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.similarity" title="Permalink to this definition">¶</a></dt>
<dd><p>Replicates the pctsimilarity field from blasr, calculated as
#matches/mean(aligned_length, read_length)</p>
</dd></dl>

<dl class="method">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.transcript">
<code class="descname">transcript</code><span class="sig-paren">(</span><em>orientation='native'</em>, <em>style='gusfield'</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>A text representation of the alignment moves (see Gusfield).
This can be useful in pretty-printing an alignment.</p>
</dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.CmpH5IO.CmpH5Alignment.zScore">
<code class="descname">zScore</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.CmpH5Alignment.zScore" title="Permalink to this definition">¶</a></dt>
<dd><p>(PacBio internal files only)</p>
<p>The z-score of the alignment, using a null model of a random
sequence alignment.</p>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="pbcore.io.align.CmpH5IO.EmptyCmpH5Error">
<em class="property">exception </em><code class="descclassname">pbcore.io.align.CmpH5IO.</code><code class="descname">EmptyCmpH5Error</code><a class="headerlink" href="#pbcore.io.align.CmpH5IO.EmptyCmpH5Error" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/exceptions.html#exceptions.Exception" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></code></a></p>
<p>An exception raised when CmpH5Reader tries to read from a
cmp.h5 with no alignments.</p>
</dd></dl>

</div>
<div class="section" id="module-pbcore.io.align.PacBioBamIndex">
<span id="pbcore-io-align-pacbiobamindex-module"></span><h2>pbcore.io.align.PacBioBamIndex module<a class="headerlink" href="#module-pbcore.io.align.PacBioBamIndex" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pbcore.io.align.PacBioBamIndex.PacBioBamIndex">
<em class="property">class </em><code class="descclassname">pbcore.io.align.PacBioBamIndex.</code><code class="descname">PacBioBamIndex</code><span class="sig-paren">(</span><em>pbiFilename</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.PacBioBamIndex.PacBioBamIndex" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>The PacBio BAM index is a companion file allowing modest
<em>semantic</em> queries on PacBio BAM files without iterating over the
entire file.  By convention, the PacBio BAM index has extension
&#8220;bam.pbi&#8221;.</p>
<dl class="attribute">
<dt id="pbcore.io.align.PacBioBamIndex.PacBioBamIndex.columnNames">
<code class="descname">columnNames</code><a class="headerlink" href="#pbcore.io.align.PacBioBamIndex.PacBioBamIndex.columnNames" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pbcore.io.align.PacBioBamIndex.PacBioBamIndex.rangeQuery">
<code class="descname">rangeQuery</code><span class="sig-paren">(</span><em>winId</em>, <em>winStart</em>, <em>winEnd</em><span class="sig-paren">)</span><a class="headerlink" href="#pbcore.io.align.PacBioBamIndex.PacBioBamIndex.rangeQuery" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pbcore.io.align.PacBioBamIndex.PacBioBamIndex.version">
<code class="descname">version</code><a class="headerlink" href="#pbcore.io.align.PacBioBamIndex.PacBioBamIndex.version" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pbcore.io.align">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pbcore.io.align" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">pbcore.io.align package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-pbcore.io.align.BamAlignment">pbcore.io.align.BamAlignment module</a></li>
<li><a class="reference internal" href="#module-pbcore.io.align.BamIO">pbcore.io.align.BamIO module</a></li>
<li><a class="reference internal" href="#module-pbcore.io.align.BlasrIO">pbcore.io.align.BlasrIO module</a></li>
<li><a class="reference internal" href="#module-pbcore.io.align.CmpH5IO">pbcore.io.align.CmpH5IO module</a></li>
<li><a class="reference internal" href="#module-pbcore.io.align.PacBioBamIndex">pbcore.io.align.PacBioBamIndex module</a></li>
<li><a class="reference internal" href="#module-pbcore.io.align">Module contents</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/pbcore.io.align.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pbcore 1.2.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011-2015, Pacific Biosciences.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>